woocommerce/woocommerce
# Checkout: Add entities for checkout fields that can be toggled on/off

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

Moves the visibility settings (hidden and required) for phone, company, and address 2 inputs to global settings instead of block attributes.

This allows the settings to be visible globally across all blocks as well as via the Store API. This approach has several benefits since we know at store level which fields are enabled. The main drawback is you can no longer have different settings per checkout block, but multiple checkout blocks is an edge use case.

Attributes from the checkout block are migrated to the global settings on access if the main checkout page is the checkout block, and the settings are blank. **We could look into migrating even if the settings are not blank, TBD**.

These global settings were already in place:

- `woocommerce_checkout_company_field`
- `woocommerce_checkout_address_2_field`
- `woocommerce_checkout_phone_field`

I believe these are used in the customizer (see `WC_Shop_Customizer::add_checkout_section` and contain values `hidden`, `required`, and `optional`. We match this in the API and the blocks.

In the editor, instead of the field settings being saved to attributes we use entities linked to wp settings. To handle the preview, this PR uses the existing `previewData` to populate the core fields combined with `getEditedEntityRecord` for said fields. On the frontend, this is not needed.

~There is once big caveat with this PR and that is the previously used attributes are discarded. Because they live at block level, I cannot see a good way to persist these globally. Therefore on upgrade, some fields on checkout may change behaviour until the merchant edits the block. This shouldn't be breaking for customers since we'll use the defaults (phone optional, address optional, company hidden).~

Closes #42773



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Edit the page containing the checkout block
2. Focus on the address block
3. Try toggling on/off phone/company/apartment fields, as well as the required/optional checkbox and ensure the preview output matches your selections
4. Save changes
5. Go to the frontend block checkout
6. Confirm the changes made in the editor are reflected on the checkout
7. Go to the frontend legacy shortcode checkout
8. The phone/address 2/company fields should also match the changes you made when editing the checkout block

To test the migration process (this is also covered by tests):

1. On a clean install setup the checkout block before using this PR. The checkout block must be on the main checkout page.
2. Change some field settings e.g. make phone required, hide address 2, etc. Save.
3. Use this PR.
4. Confirm the block settings were persisted.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
